README.md
<!-- markdownlint-disable MD032 MD033-->

# 🔥 **Josee9988/Mechanography-simulator**

<div align="center">
  <a href="https://github.com/Josee9988/Mechanography-simulator">
    <img width="45%" src="https://i.imgur.com/PuInN3X.png" alt="@Josee9988/Mechanography-simulator's logo">
  </a>
  <br>
  <a href="https://github.com/Josee9988/Mechanography-simulator/issues">
    <img src="https://img.shields.io/github/issues/Josee9988/Mechanography-simulator?color=0088ff&style=for-the-badge&logo=github" alt="Josee9988/Mechanography-simulator's issues"/>
  </a>
  <a href="https://github.com/Josee9988/Mechanography-simulator/pulls">
    <img src="https://img.shields.io/github/issues-pr/Josee9988/Mechanography-simulator?color=0088ff&style=for-the-badge&logo=github"  alt="Josee9988/Mechanography-simulator's pull requests"/>
  </a>
</div>

---

## 🤔 **About the project**

* A Python3 script that simulates human writing (on their keyboard). (control the speed, randomness, rate of typos and more!)
* The script will pretend that you are typing a text on your keyboard with random parameters and even typos to make it look more legit.
* The user can control the speed of typing, the rate of random speed of typing and the rate of typos.
* The script is cross-platform compatible with Windows, macOS and Linux.
* The script supports Spanish 🇪🇸 and English 🇬🇧🇺🇸!

---

## ⚡ **Installation**

* Go to the [**releases** section](https://github.com/Josee9988/Mechanography-simulator/releases) and download
  the latest version for your operating system.

---

## 🚀 **Usage**

* Simply execute the program and introduce the values: **rate of typos**, **speed rate**, **speed randomness added** and
  the **text** ;D

<img src="https://i.imgur.com/7iesvpS.gif" alt="execution" title="execution"/>

* The user can specify the:
  1. **Rate of typos** (integer)
  2. **Pulsation speed** (integer or decimal)
  3. **Pulsation random speed rate added** (integer or decimal)
  4. **Text** to be written

---

## 📝 **Additional notes**

* This script can be modified to fit multiple needs and can be easily added new features by the user, so feel free to fork it.

---

## 🍰 **Supporters and donators**

<!-- Change your small logo -->
<a href="https://github.com/Josee9988/Mechanography simulator">
  <img alt="@Josee9988/Mechanography simulator's brand logo without text" align="right" src="https://i.imgur.com/PuInN3X.png" width="17%" />
</a>

We are currently looking for new donators to help and maintain this project! ❤️

By donating, you will help the development of this project and *you will be featured in this Mechanography-simulator's
README.md* so, everyone can see your kindness and visit your content ⭐.

<a href="https://i.imgur.com/hnLIiDe.gif">
  <img src="https://img.shields.io/badge/Sponsor-Josee9988/Mechanography simulator-blue?logo=github-sponsors&style=for-the-badge&color=red">
</a>

<!-- LINK TO YOUR DONATING PAGES HERE -->

---

Mechanography-simulator was generated from *[Josee9988/project-template](https://github.com/Josee9988/project-template)*
📚

---

## 🎉 Was the script helpful? Help us raise these numbers up

[![GitHub's followers](https://img.shields.io/github/followers/Josee9988.svg?style=social)](https://github.com/Josee9988)
[![GitHub stars](https://img.shields.io/github/stars/Josee9988/Mechanography-simulator.svg?style=social)](https://github.com/Josee9988/Mechanography-simulator/stargazers)
[![GitHub watchers](https://img.shields.io/github/watchers/Josee9988/Mechanography-simulator.svg?style=social)](https://github.com/Josee9988/Mechanography-simulator/watchers)
[![GitHub forks](https://img.shields.io/github/forks/Josee9988/Mechanography-simulator.svg?style=social)](https://github.com/Josee9988/Mechanography-simulator/network/members)
[![Sponsor](https://img.shields.io/static/v1?label=Sponsor&message=%E2%9D%A4&logo=github-sponsors&color=red&style=social)](https://github.com/sponsors/Josee9988)

Enjoy! 😃

---

## ⚖️📝 **License and Changelog**

See the license in the '**[LICENSE](LICENSE)**' file.

Watch the changes in the '**[CHANGELOG.md](CHANGELOG.md)**' file.

---

_Made with a lot of ❤️❤️ by **[@Josee9988](https://github.com/Josee9988)**_


CHANGELOG.md
<!-- markdownlint-disable MD024-->
# **Change Log** 📜📝

All notable changes to the "**Mechanography-simulator**" script will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [**1.1.4**] - 2021-11-25

### Added

* Recommended parameters as global variables.

### Changed

* Modified values of the recommended parameters.
* From single quotes to double quotes.

### Fixed

* Minor code style changes.
* Spanish greeting message.

### Removed

* Default parameter values at the write function.

## [**1.1.3**] - 2021-11-23

### Fixed

* Some problems with the automatic sleep time.

## [**1.1.2**] - 2021-11-23

### Fixed

* Some code fixes (refactored some code and added some missing type hinting).

## [**1.1.1**] - 2021-11-18

### Fixed

* Some typos in the README.md file.
* Some minor type hinting mistakes.

## [**1.1.0**] - 2021-11-18

### Added

* Extra type hinting support

### Fixed

* The language detection for Spanish

## [**1.1.0**] - 2021-11-18

### Added

* Support for the English language.
* A file with the translations.
* A 'src/' folder to contain the files.
* When an error is written it will also write the correct character after it.
* Function type hinting.

### Removed

* Reduced import size by importing only the used functions

## [**1.0.0**] - 2021-11-17

### Added

* A decent README.md file.
* Images for the icon.
* Colours will only be added if the user is not on Windows.

### Changed

* The gitignore to ignore the build.

## [**0.0.1**] - 2021-11-17

### Added

* The basic project structure from **[josee9988/project-template](https://github.com/Josee9988/project-template)**.


src/Language_dictionary.py
from locale import getlocale

RECOMENDED_TYPOS: str = "25"
RECOMENDED_SPEED: str = "0.2"
RECOMENDED_RANDOM_SPEED_ADDED: str = "0.02"


def obtain_mechanography_language_dictionary() -> dict:
    # Obtains the system language and writes the print texts in English or Spanish. The function returns the dict
    if 'En' in getlocale()[0] or 'en' in getlocale()[0]:  # ENGLISH 🇬🇧🇺🇸
        language_dict = {"Greeting": "Script made by @Josee9988 | Jose Gracia Berenguer ;)\n\n",
                         "Text_parameter": "Introduce the text to be written: (double enter to start)\n",
                         "Error_parameter": "Introduce the rate of typos (every X +- randomness a typo will occur) ("
                                            f"eg: {RECOMENDED_TYPOS}): ",
                         "Type_rate_parameter": f"Introduce the pulsation speed (eg: {RECOMENDED_SPEED}): ",
                         "Type_rate_added_parameter": "Introduce typing rate to "
                                                      f"be added randomly (eg: {RECOMENDED_RANDOM_SPEED_ADDED}): ",
                         "Countdown": "%sPlease, point the mouse wherever you want to write, the script will begin "
                                      "in %d seconds%s",
                         "Remaining_characters": "Characters remaining: %d     ",
                         "Writing_characters": "\n%sWriting %d characters...%s\n",
                         "Press_any": "Press any key to continue... ",
                         "Restart": "\nDo you wish to restart the script (with different data)"}
    else:  # SPANISH 🇪🇸
        language_dict = {"Greeting": "Script hecho por @Josee9988 | Jose Gracia Berenguer ;)\n\n",
                         "Text_parameter": "Introduce el texto a escribir: (doble intro para validarlo)\n",
                         "Error_parameter": "Introduce cada cuántos carácteres se realizará un fallo (cada X +- un "
                                            "ratio aleatorio para que ocurra un "
                                            f"error tipográfico) (ej: {RECOMENDED_TYPOS}): ",
                         "Type_rate_parameter": f"Introduce tasa de pulsaciones (ej: {RECOMENDED_SPEED}): ",
                         "Type_rate_added_parameter": "Introduce tasa de aleatoriedad de pulsaciones a añadir (ej: "
                                                      f"{RECOMENDED_RANDOM_SPEED_ADDED}): ",
                         "Countdown": "%sPor favor, situa el ratón donde desees escribir, el programa empezará en %d "
                                      "segundos%s",
                         "Remaining_characters": "Carácteres restantes: %d     ",
                         "Writing_characters": "\n%sEscribiendo %d carácteres...%s\n",
                         "Press_any": "Pulsa cualquier tecla para continuar... ",
                         "Restart": "\nDeseas volver a ejecutar el programa (con otros datos)"}
    return language_dict


Joses_Mechanography_tester.py
#!/usr/bin/python

import pynput
import random
import time

metadata = dict(
    __title__='Jose mechanography tester',
    __author__="Jose Gracia Berenguer",
    __email__="jgracia9988@gmail.com",
    __status__="Production",
    __copyright__='MIT',
    __license__='MIT',
    __version__='1.0.0',
    __summary__='Program to write as if it were the keyboard to test mechanography programs and pretend that you are writing (super fast)',
    __uri__='https://jgracia.es',
)

globals().update(metadata)

KEYBOARD = pynput.keyboard.Controller()  # Create the controller
ERROR_LETTERS = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ` '


def write_keyboard_with_delay_and_randomness(text: str, error: int, speed: float, speed_rate_added: float) -> None:
    error_rate = error
    i = len(text)
    for character in text:  # Loop over each character in the string
        error_rate = error_rate - random.uniform(0, 1)
        if error_rate < 0:  # type an error
            KEYBOARD.type(random.choice(ERROR_LETTERS))
            error_rate = error
        else:  # Type the character
            KEYBOARD.type(character)
        time.sleep(random.uniform(0, speed) +
                   random.uniform(0, speed_rate_added))  # Sleep for the amount of seconds generated
        i = i-1
        print("Carácteres restantes: %d" % (i), end='\r')


def yes_or_no(question) -> bool:
    reply = input(question+' (s/n): ').lower().strip()
    if reply[0] == 'S' or reply[0] == 's':
        return True
    if reply[0] == 'N' or reply[0] == 'n':
        return False
    else:
        return yes_or_no("Por favor, introduce 's' o 'n''.")


def ask_and_execute() -> None:
    start_time = 5
    error_rate = int(
        input("Introduce cada cuántos carácteres se realizará un fallo: "))
    speed = float(input("Introduce tasa de pulsaciones (ej: 0.1): "))
    speed_rate_added = float(
        input("Introduce tasa de aleatoriedad de pulsaciones a añadir (ej: 0.05): "))
    print(("Introduce el texto a escribir: (doble intro para validarlo)\n"))

    text = '\n'.join(iter(input, ""))
    text_parsed = text.replace("\n", " ").strip()

    for _ in range(5):
        print("\033[1;32;40mPor favor, situa el ratón donde desees escribir, el programa empezará en %d segundos\033[0;0m" %
              start_time, end='\r')
        start_time = start_time - 1
        time.sleep(1)

    print("\n\033[1;31;40mEscribiendo %s carácteres...\033[0;0m\n" %
          len(text_parsed))

    write_keyboard_with_delay_and_randomness(
        text_parsed, error_rate, speed, speed_rate_added)  # Execute the function


### MAIN ###

print("Programa hecho por @Josee9988 \ Jose Gracia Berenguer ;)\n\n")

ask_and_execute()
input("Pulsa cualquier tecla para continuar... ")

while(yes_or_no("\nDeseas volver a ejecutar el programa (con otros datos)")):
    ask_and_execute()
    input("Pulsa cualquier tecla para continuar... ")


src/Mechanography_tester.py
#!/usr/bin/python

from os import name
from random import uniform, choice
from time import sleep
from typing import Dict, Union

from pynput import keyboard

from Language_dictionary import obtain_mechanography_language_dictionary

globals().update(dict(
    __title__="Mechanography tester",
    __author__="Jose Gracia Berenguer",
    __email__="jgracia9988@gmail.com",
    __status__="Production",
    __copyright__="Free Software License",
    __license__="MIT",
    __version__="1.1.4",
    __summary__="A Python3 script that simulates the user typing a text on their keyboard. (control the speed, "
                "randomness, rate of typos and more!)",
    __uri__="https://jgracia.es",
))  # metadata

KEYBOARD: any = keyboard.Controller()  # Create the controller
LANGUAGE_OUTPUT: dict = obtain_mechanography_language_dictionary()
COUNTDOWN_TIME: int = 5  # seconds
ERROR_LETTERS: str = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ` "
COLOUR_RED: str = "\033[1;31;40m"
COLOUR_GREEN: str = "\033[1;32;40m"
COLOUR_RESET: str = "\033[0;0m"

if name == "nt":  # If we are running on Windows remove all the colours
    COLOUR_RED = COLOUR_GREEN = COLOUR_RESET = ""


def write_as_keyboard(text: str, error: int, speed: float, speed_rate_added: float) -> None:
    error_rate: float = error
    i: int = len(text)
    for character in text:  # Loop over each character in the string
        error_rate = error_rate - uniform(0, 1)
        if error_rate < 0:  # type an error
            KEYBOARD.type(choice(ERROR_LETTERS))
            KEYBOARD.type(character)
            error_rate = error
        else:  # Type the character
            KEYBOARD.type(character)
        sleep(speed + uniform(0, speed_rate_added))  # Sleep
        i = i - 1
        print(LANGUAGE_OUTPUT["Remaining_characters"] % i, end="\r")


def yes_or_no(question) -> bool:
    reply: str = input(question + " yes/no (y/n): ").lower().strip()
    if reply[0] == "Y" or reply[0] == "y":
        return True
    if reply[0] == "N" or reply[0] == "n":
        return False
    else:
        return yes_or_no("Introduce 'Y' or 'N'.")


# returns text, error, speed and speed_rate
def obtain_parameters() -> Dict[str, Union[int, float, str]]:
    parameters: dict = {"Text": "", "Error": int(input(LANGUAGE_OUTPUT["Error_parameter"])),
                        "Speed": float(input(LANGUAGE_OUTPUT["Type_rate_parameter"])),
                        "Speed_rate_added": float(input(LANGUAGE_OUTPUT["Type_rate_added_parameter"]))}

    print(LANGUAGE_OUTPUT["Text_parameter"])  # Ask for the text
    text: str = "\n".join(iter(input, ""))
    parameters["Text"] = text.replace("\n", " ").strip()
    return parameters


def announce_countdown(text: str, countdown_duration: int) -> None:
    for _ in range(5):
        print(LANGUAGE_OUTPUT["Countdown"] % (
            COLOUR_GREEN, countdown_duration, COLOUR_RESET), end="\r")
        countdown_duration = countdown_duration - 1
        sleep(1)

    print(LANGUAGE_OUTPUT["Writing_characters"] %
          (COLOUR_RED, len(text), COLOUR_RESET))


def main_handler() -> None:
    parameters: dict = obtain_parameters()
    announce_countdown(parameters["Text"], COUNTDOWN_TIME)
    write_as_keyboard(parameters["Text"], parameters["Error"],
                      parameters["Speed"], parameters["Speed_rate_added"])
    input(LANGUAGE_OUTPUT["Press_any"])


# -- MAIN -- #
print(LANGUAGE_OUTPUT["Greeting"])
main_handler()

while yes_or_no(LANGUAGE_OUTPUT["Restart"]):
    main_handler()
